File: packages/config-mutations-rust/src/execution.js
import {applyConfig} from './config.js';
import {isNotFound} from './fs-utils.js';
import {applyTemplate} from './template.js';
import {applyBackup} from './backup.js';
import {writeWalk} from './path.js';
import path from 'node:path';
import {createRequire} from 'node:module';
const native=createRequire(import.meta.url)('./config-mutations-rust.node');
function factories(specifications){return Object.fromEntries(specifications.map(({name,kind,fields})=>[name,options=>Object.fromEntries([['kind',kind],...fields.map(field=>[field,options[field]])])]));}
const layouts=native.configMutationFactories(),knownKinds=new Set(Object.values(layouts).flatMap(group=>group.map(factory=>factory.kind)));
export const fileMutation=factories(layouts.file);
export const configMutation=factories(layouts.config);
export const templateMutation=factories(layouts.template);
function expandHome(target,home){
 if(target.startsWith('~./'))target=`~/.${target.slice(3)}`;
 let remainder=target.slice(1);
 if(remainder.startsWith('/')||remainder.startsWith('\\'))remainder=remainder.slice(1);
 else if(remainder.startsWith('.')){remainder=remainder.slice(1);if(remainder.startsWith('/')||remainder.startsWith('\\'))remainder=remainder.slice(1);}
 return remainder.length===0?home:path.join(home,remainder);
}
function resolvePath(raw,context){
 if(typeof raw!=='string'||raw.length===0)throw Error('Target path must be a non-empty string.');
 if(!raw.startsWith('~'))throw Error(`All target paths must be home-relative (start with ~). Received: "${raw}"`);
 const expanded=expandHome(raw,context.homeDir),canonical=path.resolve(expanded),relative=path.relative(path.resolve(context.homeDir),canonical);
 if(relative==='..'||relative.startsWith(`..${path.sep}`)||path.isAbsolute(relative))throw Error(`Target path resolves outside home directory: "${raw}"`);
 if(!context.pathMapper)return canonical;
 const directory=context.pathMapper.mapTargetDirectory({targetDirectory:path.dirname(expanded)}),filename=path.basename(expanded);
 return filename.length===0?directory:path.join(directory,filename);
}
function resolveTarget(mutation,options){const kind=mutation.kind;if(!knownKinds.has(kind))return undefined;const value=['ensureDirectory','removeDirectory'].includes(kind)?mutation.path:mutation.target;return typeof value==='function'?value(options):value;}
function label(kind,target){const display=target??'target';switch(kind){case 'ensureDirectory':return `Create ${display}`;case 'removeDirectory':return `Remove directory ${display}`;case 'removeFile':return `Remove ${display}`;case 'chmod':return `Set permissions on ${display}`;case 'backup':return `Backup ${display}`;case 'restoreBackup':return `Restore ${display}`;case 'configMerge':case 'configPrune':case 'configTransform':case 'templateMergeJson':case 'templateMergeToml':return `Update ${display}`;case 'templateWrite':return `Write ${display}`;default:return 'Operation';}}
function pendingDetails(mutation,context,options){
 try{const raw=resolveTarget(mutation,options);if(raw===undefined)return {kind:mutation.kind,label:mutation.label??mutation.kind};
  try{const targetPath=resolvePath(raw,context);return {kind:mutation.kind,label:mutation.label??label(mutation.kind,targetPath),targetPath};}
  catch{return {kind:mutation.kind,label:mutation.label??label(mutation.kind,raw),targetPath:undefined};}
 }catch{return {kind:mutation.kind,label:mutation.label??mutation.kind};}
}

function prepareTarget(mutation,context,options){const raw=resolveTarget(mutation,options),target=resolvePath(raw,context);return {raw,target,details:{kind:mutation.kind,label:mutation.label??label(mutation.kind,target),targetPath:target}};}
async function applyFile(mutation,context,options){
 if(!knownKinds.has(mutation.kind))throw Error(`Unknown mutation kind: ${mutation.kind}`);
 if(['templateWrite','templateMergeJson','templateMergeToml'].includes(mutation.kind))return applyTemplate(mutation,context,options,()=>prepareTarget(mutation,context,options));
 const {raw,target:targetPath,details}=prepareTarget(mutation,context,options);
 if(['configMerge','configPrune','configTransform'].includes(mutation.kind))return {outcome:await applyConfig(mutation,context,options,raw,targetPath),details};
 if(mutation.kind==='backup'||mutation.kind==='restoreBackup')return {outcome:await applyBackup(mutation,context,targetPath),details};
 const machine=new native.ConfigFileMachine(mutation.kind,writeWalk(targetPath,context.homeDir));
 let request=machine.start();
 while(true){
  if(request.error)throw Error(request.error);
  if(request.kind==='done')return {outcome:request.outcome,details};
  let response;
  try{switch(request.kind){
   case 'chmodSupported':response={kind:'chmodSupported',flag:typeof context.fs.chmod==='function'};break;
   case 'directoryOptions':{const supported=typeof context.fs.rm==='function';response={kind:'directoryOptions',flag:supported,force:supported?!!mutation.force:false};break;}
   case 'mode':response={kind:'mode',mode:mutation.mode};break;
   case 'dryRun':response={kind:'dryRun',flag:!!context.dryRun};break;
   case 'lstat':response={kind:'link',flag:(await context.fs.lstat(request.path)).isSymbolicLink()};break;
   case 'stat':{const stat=await context.fs.stat(targetPath);response={kind:'stat',mode:typeof stat.mode==='number'?stat.mode:undefined};break;}
   case 'readFile':response={kind:'content',content:await context.fs.readFile(targetPath,'utf8')};break;
   case 'readdir':response={kind:'count',count:(await context.fs.readdir(targetPath)).length};break;
   case 'guard':{if(mutation.whenContentMatches)mutation.whenContentMatches.lastIndex=0;const matches=!mutation.whenContentMatches||mutation.whenContentMatches.test(request.content);response={kind:'guard',flag:matches,whenEmpty:matches?!!mutation.whenEmpty:false};break;}
   case 'mkdir':await context.fs.mkdir(targetPath,{recursive:true});response={kind:'unit'};break;
   case 'rm':await context.fs.rm(targetPath,{recursive:true,force:true});response={kind:'unit'};break;
   case 'unlink':await context.fs.unlink(targetPath);response={kind:'unit'};break;
   case 'chmod':await context.fs.chmod(targetPath,mutation.mode);response={kind:'unit'};break;
   default:throw Error(`Unknown native filesystem request: ${request.kind}`);
  }}catch(error){
   if(!isNotFound(error))throw error;
   if(request.kind==='lstat'||request.kind==='stat'||mutation.kind==='removeFile'||mutation.kind==='chmod')response={kind:'missing'};else throw error;
  }
  request=machine.respond(response);
 }
}
/** Execute supported file and configuration mutations in order with an injected filesystem. */
export async function runMutations(mutations,context,options){
 const effects=[],resolverOptions=options??{};let changed=false;
 for(const mutation of mutations){const pending=pendingDetails(mutation,context,resolverOptions);context.observers?.onStart?.(pending);
  try{const {outcome,details}=await applyFile(mutation,context,resolverOptions);context.observers?.onComplete?.(details,outcome);effects.push(outcome);if(outcome.changed)changed=true;}
  catch(error){context.observers?.onError?.(pending,error);throw error;}
 }
 return {changed,effects};
}

File: packages/config-mutations-rust/src/execution.rs
//! Dependency-injected file mutation state machine. The caller executes only
//! requested effects and reads host controls when requested, preserving lazy
//! getters and mutations across awaited I/O without retaining a foreign runtime.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Kind {
    EnsureDirectory,
    RemoveDirectory,
    RemoveFile,
    Chmod,
}
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Outcome {
    pub changed: bool,
    pub effect: &'static str,
    pub detail: &'static str,
}
impl Outcome {
    pub fn noop() -> Self {
        Self {
            changed: false,
            effect: "none",
            detail: "noop",
        }
    }
    fn deleted() -> Self {
        Self {
            changed: true,
            effect: "delete",
            detail: "delete",
        }
    }
}
#[derive(Clone, Debug, PartialEq)]
pub enum Request {
    ChmodSupported,
    DirectoryOptions,
    Mode,
    DryRun,
    InspectLink(Vec<u16>),
    Stat,
    ReadFile,
    ReadEntries,
    Guard(Vec<u16>),
    MakeDirectory,
    RemoveDirectory,
    Unlink,
    SetMode(f64),
    Done(Outcome),
}
#[derive(Clone, Debug)]
pub enum Response {
    Unit,
    Missing,
    Link(bool),
    Stat(Option<f64>),
    Content(Vec<u16>),
    Count(usize),
    ChmodSupported(bool),
    DirectoryOptions { supported: bool, force: bool },
    Mode(f64),
    Guard { matches: bool, when_empty: bool },
    DryRun(bool),
}
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
enum Stage {
    New,
    ChmodSupported,
    DirectoryOptions,
    Mode,
    DryRun,
    Link,
    Stat,
    ReadFile,
    ReadEntries,
    Guard,
    Effect,
    Done,
}
pub struct FileMachine {
    kind: Kind,
    write_walk: Vec<Vec<u16>>,
    link_index: usize,
    stage: Stage,
    outcome: Option<Outcome>,
    pending_effect: Option<Request>,
    content_empty: bool,
    current_mode: Option<f64>,
}
fn error(message: &str) -> Vec<u16> {
    message.encode_utf16().collect()
}
impl FileMachine {
    pub fn new(kind: Kind, write_walk: Vec<Vec<u16>>) -> Self {
        Self {
            kind,
            write_walk,
            link_index: 0,
            stage: Stage::New,
            outcome: None,
            pending_effect: None,
            content_empty: false,
            current_mode: None,
        }
    }
    pub fn start(&mut self) -> Result<Request, Vec<u16>> {
        if self.stage != Stage::New {
            return Err(error("File mutation already started"));
        }
        if self.kind == Kind::Chmod {
            self.stage = Stage::ChmodSupported;
            return Ok(Request::ChmodSupported);
        }
        Ok(self.begin_walk())
    }
    fn begin_walk(&mut self) -> Request {
        if matches!(self.kind, Kind::EnsureDirectory | Kind::Chmod) && !self.write_walk.is_empty() {
            self.stage = Stage::Link;
            Request::InspectLink(self.write_walk[0].clone())
        } else {
            self.begin_inspection()
        }
    }
    fn begin_inspection(&mut self) -> Request {
        if self.kind == Kind::RemoveFile {
            self.stage = Stage::ReadFile;
            Request::ReadFile
        } else {
            self.stage = Stage::Stat;
            Request::Stat
        }
    }
    fn done(&mut self, outcome: Outcome) -> Request {
        self.stage = Stage::Done;
        self.write_walk = vec![];
        Request::Done(outcome)
    }
    fn effect(&mut self, request: Request, outcome: Outcome) -> Request {
        self.stage = Stage::DryRun;
        self.outcome = Some(outcome);
        self.pending_effect = Some(request);
        Request::DryRun
    }
    pub fn respond(&mut self, response: Response) -> Result<Request, Vec<u16>> {
        match (self.stage, response) {
            (Stage::ChmodSupported, Response::ChmodSupported(supported)) => {
                if supported {
                    Ok(self.begin_walk())
                } else {
                    Ok(self.done(Outcome::noop()))
                }
            }
            (Stage::Link, Response::Link(true)) => {
                let mut message = error("Refusing mutation write through symbolic link: ");
                message.extend_from_slice(&self.write_walk[self.link_index]);
                self.stage = Stage::Done;
                self.write_walk = vec![];
                Err(message)
            }
            (Stage::Link, Response::Link(false) | Response::Missing) => {
                self.link_index += 1;
                if let Some(path) = self.write_walk.get(self.link_index) {
                    Ok(Request::InspectLink(path.clone()))
                } else {
                    Ok(self.begin_inspection())
                }
            }
            (Stage::Stat, Response::Missing) => {
                if self.kind == Kind::EnsureDirectory {
                    Ok(self.effect(
                        Request::MakeDirectory,
                        Outcome {
                            changed: true,
                            effect: "mkdir",
                            detail: "create",
                        },
                    ))
                } else {
                    Ok(self.done(Outcome::noop()))
                }
            }
            (Stage::Stat, Response::Stat(mode)) => match self.kind {
                Kind::EnsureDirectory => Ok(self.effect(
                    Request::MakeDirectory,
                    Outcome {
                        changed: false,
                        effect: "mkdir",
                        detail: "noop",
                    },
                )),
                Kind::RemoveDirectory => {
                    self.stage = Stage::DirectoryOptions;
                    Ok(Request::DirectoryOptions)
                }
                Kind::Chmod => {
                    self.current_mode = mode.map(|n| {
                        if n.is_finite() {
                            (n.trunc().rem_euclid(4294967296.0) as u32 & 0o777) as f64
                        } else {
                            0.0
                        }
                    });
                    self.stage = Stage::Mode;
                    Ok(Request::Mode)
                }
                Kind::RemoveFile => unreachable!(),
            },
            (Stage::DirectoryOptions, Response::DirectoryOptions { supported, force }) => {
                if !supported {
                    Ok(self.done(Outcome::noop()))
                } else if force {
                    Ok(self.effect(Request::RemoveDirectory, Outcome::deleted()))
                } else {
                    self.stage = Stage::ReadEntries;
                    Ok(Request::ReadEntries)
                }
            }
            (Stage::Mode, Response::Mode(mode)) => {
                if self.current_mode == Some(mode) {
                    Ok(self.done(Outcome::noop()))
                } else {
                    Ok(self.effect(
                        Request::SetMode(mode),
                        Outcome {
                            changed: true,
                            effect: "chmod",
                            detail: "update",
                        },
                    ))
                }
            }
            (Stage::ReadEntries, Response::Count(count)) => {
                if count > 0 {
                    Ok(self.done(Outcome::noop()))
                } else {
                    Ok(self.effect(Request::RemoveDirectory, Outcome::deleted()))
                }
            }
            (Stage::ReadFile, Response::Content(content)) => {
                let content = trim(&content).to_vec();
                self.content_empty = content.is_empty();
                self.stage = Stage::Guard;
                Ok(Request::Guard(content))
            }
            (Stage::ReadFile | Stage::Guard | Stage::Mode | Stage::DryRun, Response::Missing)
                if matches!(self.kind, Kind::RemoveFile | Kind::Chmod) =>
            {
                Ok(self.done(Outcome::noop()))
            }
            (
                Stage::Guard,
                Response::Guard {
                    matches,
                    when_empty,
                },
            ) => {
                if !matches || when_empty && !self.content_empty {
                    Ok(self.done(Outcome::noop()))
                } else {
                    Ok(self.effect(Request::Unlink, Outcome::deleted()))
                }
            }
            (Stage::DryRun, Response::DryRun(dry_run)) => {
                if dry_run {
                    self.pending_effect.take();
                    let outcome = self.outcome.take().unwrap();
                    Ok(self.done(outcome))
                } else {
                    self.stage = Stage::Effect;
                    Ok(self.pending_effect.take().unwrap())
                }
            }
            (Stage::Effect, Response::Unit) => {
                let outcome = self.outcome.take().unwrap();
                Ok(self.done(outcome))
            }
            (Stage::Effect, Response::Missing)
                if matches!(self.kind, Kind::RemoveFile | Kind::Chmod) =>
            {
                Ok(self.done(Outcome::noop()))
            }
            _ => Err(error("Unexpected file mutation response")),
        }
    }
}
/// ECMAScript trim removes whitespace and line terminators, including BOM,
/// but not zero-width space or the former Mongolian vowel separator.
fn trim(source: &[u16]) -> &[u16] {
    fn whitespace(unit: u16) -> bool {
        matches!(unit,0x0009..=0x000d|0x0020|0x00a0|0x1680|0x2000..=0x200a|0x2028|0x2029|0x202f|0x205f|0x3000|0xfeff)
    }
    let start = source
        .iter()
        .position(|u| !whitespace(*u))
        .unwrap_or(source.len());
    let end = source
        .iter()
        .rposition(|u| !whitespace(*u))
        .map_or(start, |n| n + 1);
    &source[start..end]
}

/// Declarative host factory layouts preserve foreign resolver/guard identities.
pub struct Factory {
    pub name: &'static str,
    pub kind: &'static str,
    pub fields: &'static [&'static str],
}
pub const FILE_FACTORIES: &[Factory] = &[
    Factory {
        name: "ensureDirectory",
        kind: "ensureDirectory",
        fields: &["path", "label"],
    },
    Factory {
        name: "remove",
        kind: "removeFile",
        fields: &["target", "whenEmpty", "whenContentMatches", "label"],
    },
    Factory {
        name: "removeDirectory",
        kind: "removeDirectory",
        fields: &["path", "force", "label"],
    },
    Factory {
        name: "chmod",
        kind: "chmod",
        fields: &["target", "mode", "label"],
    },
    Factory {
        name: "backup",
        kind: "backup",
        fields: &["target", "once", "label"],
    },
    Factory {
        name: "restoreBackup",
        kind: "restoreBackup",
        fields: &["target", "label"],
    },
];

pub const CONFIG_FACTORIES: &[Factory] = &[
    Factory {
        name: "merge",
        kind: "configMerge",
        fields: &["target", "value", "format", "pruneByPrefix", "label"],
    },
    Factory {
        name: "prune",
        kind: "configPrune",
        fields: &["target", "shape", "format", "onlyIf", "label"],
    },
    Factory {
        name: "transform",
        kind: "configTransform",
        fields: &["target", "format", "transform", "label"],
    },
];
